### PR TITLE
Bug: Find command doesn't work when no attribute specified e.g (find john)

### PR DESCRIPTION
**Pull request by bgopi23**
Resolves https://github.com/AY2324S2-CS2103T-T17-3/tp/issues/149
<hr>

Before:
Screenshot 2024-03-24 at 9 59 11 AM
<hr>

After:
Screenshot 2024-03-24 at 9 58 21 AM

### CODE REVIEW

File: src/main/java/seedu/address/logic/parser/FindCommandParser.java [hunk 1]
```diff
@@ -40,23 +40,34 @@ public FindCommand parse(String args) throws ParseException {
...
-
-        CombinedPredicates predicates = new CombinedPredicates(namePredicate, phonePredicate, emailPredicate,
-                addressPredicate, notePredicate, tagsPredicate);
-
-        return new FindCommand(predicates);
+        // Check if no attribute is provided e.g (find john)
+        if (!argMultimap.getPreamble().isEmpty() && argMultimap.getValue(PREFIX_NAME).isEmpty()
+            && argMultimap.getValue(PREFIX_PHONE).isEmpty() && argMultimap.getValue(PREFIX_EMAIL).isEmpty()
+                && argMultimap.getValue(PREFIX_ADDRESS).isEmpty() && argMultimap.getValue(PREFIX_NOTE).isEmpty()
+                && argMultimap.getAllValues(PREFIX_TAG).isEmpty()) {
```
Comment (dillontkh) [inline src/main/java/seedu/address/logic/parser/FindCommandParser.java:47]: Is there a more elegant way to do this?

Perhaps a `noAttributesSpecified(): boolean` method in `ArgumentMultimap`?

Side note: `ArgumentMultimap` contains a `contains(Prefix): boolean` method that checks if a certain prefix has been provided, which seems like what you were trying to accomplish with the all the `getValue.isEmpty()` checks. Was there a reason why `contains()` couldn't be used?

Comment (bgopi23) [reply src/main/java/seedu/address/logic/parser/FindCommandParser.java]: i overlook that contains method, thanks for pointing it out, have made the changes

**Review by dillontkh — COMMENTED** [review]
Nice work, the checking of prefixes looks a lot cleaner now!

However, I noticed another possible bug.

image
image

Since the 'default' search field is name, I anticipate a scenario where users omit the `n/` tag out of habit, which might lead to inaccurate search results when they try to search using name + another field.

I think a possible solution would be to set the name predicate to the preamble if PREFIX_NAME is missing.

**Comment by bgopi23** [comment]
great catch @dillontkh 
resolved https://github.com/AY2324S2-CS2103T-T17-3/tp/pull/150#pullrequestreview-1956511929

File: src/main/java/seedu/address/logic/parser/ArgumentMultimap.java [hunk 2]
```diff
@@ -84,11 +84,20 @@ public boolean contains(Prefix prefix) {
...
+    public boolean isAnyPrefixPresent(Prefix... prefixes) {
+        return Stream.of(prefixes).anyMatch(prefix -> contains(prefix));
+    }
+
+    /**
+     * Checks if all prefixes specified is present in the ArgumentMultimap object
+     * @param prefixes to check
+     * @return true if prefix/s present and false otherwise
+     */
+    public boolean isAllPrefixesPresent(Prefix... prefixes) {
```
Comment (dillontkh) [inline src/main/java/seedu/address/logic/parser/ArgumentMultimap.java:100]: Is this method still in use?

**Review by dillontkh — DISMISSED** [review]
Thanks for the quick fix!

Apart from a duplicate method that isn't used, LGTM!

image

File: src/main/java/seedu/address/logic/parser/ArgumentMultimap.java (hunk 2 above)
Comment (bgopi23) [reply src/main/java/seedu/address/logic/parser/ArgumentMultimap.java:100]: its not duplicate, it checks if all prefixes are in argMultimap (unlike the other one which checks if any of the prefixes is in argMultimap), will be useful for some bug fixes im working on

**Comment by bgopi23** [comment]
@dillontkh  it looks very similar but its not duplicate, you can refer to https://github.com/AY2324S2-CS2103T-T17-3/tp/pull/150#discussion_r1536855495

File: src/main/java/seedu/address/logic/parser/FindCommandParser.java [hunk 3]
```diff
@@ -40,23 +41,31 @@ public FindCommand parse(String args) throws ParseException {
...
-                addressPredicate, notePredicate, tagsPredicate);
-
-        return new FindCommand(predicates);
+        // Check if no attribute is provided e.g (find john)
+        if (!argMultimap.getPreamble().isEmpty() && !argMultimap.isAnyPrefixPresent(ALL_PREFIXES)) {
+            NameContainsSubstringPredicate derivedNamePredicate = new NameContainsSubstringPredicate(
+                    ParserUtil.parseSearchString(argMultimap.getPreamble()));
+            CombinedPredicates predicates = new CombinedPredicates(derivedNamePredicate);
+            return new FindCommand(predicates);
+        } else {
```
[the remark this answers is not in the document]
Comment (bgopi23) [reply src/main/java/seedu/address/logic/parser/FindCommandParser.java]: good catch @jjchee77 the if statement was added as line 52 used to be, i didnt realise the if statement was unnecessary if i changed it to `.orElse(argMultimap.getPreamble())`

`ParserUtil.parseSearchString(argMultimap.getValue(PREFIX_NAME).orElse("")));`

[the remark this answers is not in the document]
Comment (jjchee77) [reply src/main/java/seedu/address/logic/parser/FindCommandParser.java]: Great work on quick fix!

File: src/main/java/seedu/address/logic/parser/CliSyntax.java [hunk 4]
```diff
@@ -13,4 +13,12 @@ public class CliSyntax {
...
     public static final Prefix PREFIX_NOTE = new Prefix("note/");
     public static final Prefix PREFIX_CLEAR_CONFIRM = new Prefix("/confirm");
+    public static final Prefix[] ALL_PREFIXES = {
+        PREFIX_NAME,
+        PREFIX_PHONE,
+        PREFIX_EMAIL,
+        PREFIX_ADDRESS,
+        PREFIX_TAG,
+        PREFIX_NOTE
+    };
```
Comment (jjchee77) [inline src/main/java/seedu/address/logic/parser/CliSyntax.java:23]: Just have to clear up unused code such as this one.

File: src/main/java/seedu/address/logic/parser/ArgumentMultimap.java [hunk 5]
```diff
@@ -82,4 +82,22 @@ public void verifyNoDuplicatePrefixesFor(Prefix... prefixes) throws ParseExcepti
...
```
Comment (jjchee77) [inline src/main/java/seedu/address/logic/parser/ArgumentMultimap.java:100]: Just have to clear up unused code such as this one.

**Review by jjchee77 — APPROVED** [review]
Otherwise, LGTM!

**Review by dillontkh — APPROVED** [review]
Works great! LGTM!

**Pull request merged**